### PR TITLE
docs: clean up model docs

### PR DESCRIPTION
**Pull request by ephys**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [ ] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

This is a bit of a miscellaneous PR.
What it does is:

- Fix many documentation issues in models
- Update missing typings related to models
- Synchronize model.d.ts and model.js, in preparation for their future merge
- Remove the deprecated-in-v6 `ModelCtor` and `ModelType` types
- Deprecate `Model.dropSchema` (it's completely unrelated to model, and is just an alias for `Sequelize.dropSchema` and `QueryInterface.dropSchema`)
- Some elements of the documentation (such as explaining querying in details) have been removed in favor of a link to the relevant documentation page
- Fix code examples not being wrapped in markdown code blocks
- `warnOnInvalidOptions` has been renamed to `_warnOnInvalidOptions` because it's a private method (and may be moved to model-internals in the future)

### CODE REVIEW

File: src/associations/belongs-to.d.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import { DataType } from '../data-types';
-import { CreateOptions, CreationAttributes, FindOptions, Model, ModelCtor, SaveOptions } from '../model';
+import { CreateOptions, CreationAttributes, FindOptions, Model, ModelStatic, SaveOptions } from '../model';
 import { Association, AssociationOptions, SingleAssociationAccessors } from './base';
 
-// type ModelCtor<M extends Model> = InstanceType<typeof M>;
+// type ModelStatic<M extends Model> = InstanceType<typeof M>;
```
Comment (sdepold) [inline src/associations/belongs-to.d.ts:5]: Do we need this?

File: src/dialects/abstract/query-interface.d.ts [hunk 2]
```diff
@@ -175,23 +201,21 @@ export interface IndexesOptions {
...
 
   /**
-   * An array of the fields to index. Each field can either be a string containing the name of the field,
-   * a sequelize object (e.g `sequelize.fn`), or an object with the following attributes: `name`
-   * (field name), `length` (create a prefix index of length chars), `order` (the direction the column
-   * should be sorted in), `collate` (the collation (sort order) for the column), `operator` (likes IndexesOptions['operator'])
+   * The fields to index.
    */
-  fields?: (string | { name: string; length?: number; order?: 'ASC' | 'DESC'; collate?: string; operator?: string } | Fn | Literal)[];
+  fields?: Array<string | IndexField | Fn | Literal>;
```
Comment (sdepold) [inline src/dialects/abstract/query-interface.d.ts:213]: Good stuff

File: src/model.d.ts [hunk 3]
```diff
@@ -617,23 +665,32 @@ export type WhereAttributeHashValue<AttributeType> =
...
+   * The alias for the join model, in case you want to give it a different name than the default one.
    */
   as?: string;
 
   /**
    * If true, only non-deleted records will be returned from the join table.
    * If false, both deleted and non-deleted records will be returned.
    * Only applies if through model is paranoid.
+   *
+   * @default true
```
Comment (sdepold) [inline src/model.d.ts:677]: Isn't it by default turned off?

File: src/model.d.ts [hunk 4]
```diff
@@ -1779,15 +1976,21 @@ export interface ModelOptions<M extends Model = Model> {
...
+   * Hooks will be called before and after certain operations.
+   *
+   * This can also be done through {@link Model.addHook}, or the individual hook methods such as {@link Model.afterQuery}.
+   * Each property can either be a function, or an array of functions.
+   *
+   * @see https://sequelize.org/docs/v7/other-topics/hooks/
    */
-  hooks?: Partial<ModelHooks<M, Attributes<M>>>;
+  hooks?: {
+    [Key in keyof ModelHooks<M, Attributes<M>>]?: AllowArray<ModelHooks<M, Attributes<M>>[Key]>
```
Comment (sdepold) [inline src/model.d.ts:1988]: Ooooph

**Review by sdepold — COMMENTED** [review]
First half reviewed

File: src/model.d.ts (hunk 3 above)
Comment (ephys) [reply src/model.d.ts:677]: This option is on by default but it also needs paranoid to be on to do anything

File: src/associations/belongs-to.d.ts (hunk 1 above)
Comment (ephys) [reply src/associations/belongs-to.d.ts:5]: Nope, but it's being removed in https://github.com/sequelize/sequelize/pull/14280 :)

**Comment by ephys** [comment]
Pending question: Model has both methods `setAttributes` and `set`. `setAttributes` is just an alias for `set`.

It looks like `setAttributes` was added by https://github.com/sequelize/sequelize/pull/5924

Which one should we deprecate?

In favor of keeping `setAttributes`:

- The name is more explicit than `set`, especially since we have `setDataValue` too. Could help with readability as other methods expose a `set` method with completely different behaviors.

Against keeping `setAttributes`:

- We need to add `getAttributes`, `getAttribute`, `setAttribute` to be complete
- `Model.getAttributes` and `Model#getAttributes` would have completely different meanings

That last point makes me think we should keep `set()` instead.

**Comment by WikiRik** [comment]
I'll let @sdepold review the rest


About `setAttributes`, let's do that in a future PR. It would not really be useful if `Model.getAttributes` and `Model#getAttributes` do 2 completely separate things, but better naming is also nice to have.

**Comment by ephys** [comment]
Let me know when you're ready to review, I'll resolve the conflicts then

**Review by sdepold — DISMISSED** [review]

**Comment by WikiRik** [comment]
Okay, either I resolved the conflicts wrong or we need to update something to pass the tests when merging main in this. Either way, I'll leave this up to @ephys before I break anything

**Comment by ephys** [comment]
I'll do the merge again, it looks like a few things were merged incorrectly :)
Not surprising to be honest, the scope of the changes were going to lead to complex conflicts

**Review by WikiRik — DISMISSED** [review]

**Comment by ephys** [comment]
Fixed a failing test

Scrolling through the changes, I think we're good. Let's see what the integration tests have to say

**Review by WikiRik — APPROVED** [review]

**Pull request merged**